### PR TITLE
AMPI: use zero-copy direct API to avoid intermedate copies

### PR DESCRIPTION
**Pull request by stwhite91**
*Original date: 2019-05-03 17:45:18*
*Original PR: https://charm.cs.illinois.edu/gerrit/3080*

---

For messages sent within the same PE or process, we pack an
AmpiNcpyShmBuffer object into the AmpiMsg data. This allows direct
handling of DDTs in a single message, and avoids the cost of memory
pinning when it is unnecessary. We still have to handle the case where
the recv'er turns out to not be in the same process, so the recv'er
then requests a put() of the message payload.

For messages sent across processes, we pack a CkNcpyBuffer object into
the AmpiMsg data and use that to do a get() from the recv'er.

Change-Id: I9e8126bc8bf321b4a5ab64e08b3283e448fec4c0

### CODE REVIEW

File: src/libs/ck-libs/ampi/ampiimpl.h [hunk 1]
```diff
@@ -2497,8 +2594,8 @@ class ampi final : public CBase_ampi {
   ampi() noexcept;
   ampi(CkArrayID parent_,const ampiCommStruct &s) noexcept;
   ampi(CkMigrateMessage *msg) noexcept;
-  void ckJustMigrated() noexcept;
-  void ckJustRestored() noexcept;
+  void ckJustMigrated(void) noexcept;
+  void ckJustRestored(void) noexcept;
```
Comment (evan-charmworks) [inline src/libs/ck-libs/ampi/ampiimpl.h]: Don't need these voids

Comment (stwhite91) [reply src/libs/ck-libs/ampi/ampiimpl.h]: Done

**Comment by stwhite91** [comment]
This patches fixes the failures in tests/ampi/datatype that have been silently happening for a couple months

**Comment by stwhite91** [comment]
Needs another review

File: src/libs/ck-libs/ampi/ampi.C [hunk 2]
```diff
@@ -6092,7 +6195,7 @@ AMPI_API_IMPL(int, MPI_Wait, MPI_Request *request, MPI_Status *sts)
              *request, reqs[*request], (int)(reqs[*request]->tag));
   AMPI_DEBUG("MPI_Wait: request=%d, reqs.size=%d, &reqs=%d\n",
              *request, reqs.size(), reqs);
```
Comment (matthiasdiener) [inline src/libs/ck-libs/ampi/ampi.C]: `reqs.size()` probably won't compile (see my previous patch to restore AMPI_DEBUG)

**Comment by evan-charmworks** [comment]
Still has a conflict

File: src/libs/ck-libs/ampi/ampi.C (hunk 2 above)
Comment (stwhite91) [reply src/libs/ck-libs/ampi/ampi.C]: Fixed

**Pull request merged**